ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ExternalTableCopyTaskBuilder.java

```diff
@@ -54,14 +59,53 @@
   List<Task<? extends Serializable>> tasks(TaskTracker tracker) {
     List<Task<? extends Serializable>> tasks = new ArrayList<>();
     Iterator<DirCopyWork> itr = work.getPathsToCopyIterator();
-    while (tracker.canAddMoreTasks() && itr.hasNext()) {
+    int numTaskCanBeAdded = tracker.numTaskCanBeAdded();
+    Task<? extends Serializable> barrierTask = TaskFactory.get(new DependencyCollectionWork(), conf);
+    while (numTaskCanBeAdded-- > 0 && itr.hasNext()) {
       DirCopyWork dirCopyWork = itr.next();
       Task<DirCopyWork> task = TaskFactory.get(dirCopyWork, conf);
       tasks.add(task);
-      tracker.addTask(task);
+      barrierTask.addDependentTask(task);
       LOG.debug("added task for {}", dirCopyWork);
     }
-    return tasks;
+
+    if (!tasks.isEmpty()) {
+      tracker.addDependentTask(barrierTask);
+      tracker.addTaskList(tasks);
+      return Collections.singletonList(barrierTask);
+    } else {
+      return tasks;
+    }
+  }
+
+  private static Integer setTargetPathOwnerInt(Path targetPath, Path sourcePath, HiveConf conf) throws IOException {
+    FileSystem targetFs = targetPath.getFileSystem(conf);
+    if (!targetFs.exists(targetPath)) {
+      targetFs.create(targetPath);
+    }
+    FileStatus status = sourcePath.getFileSystem(conf).getFileStatus(sourcePath);
+    if (status == null) {
+      throw new IOException("source path missing " + sourcePath);
```

Comment:
this would throw exception if the source path will not be available during load if the table was dropped in source, should handle that case, the owner should always be the one at the directory level for table /  partition

Reply:
same issue can happen for distcp done below  also ?